compiler/rustc_ty_utils/src/opaque_types.rs

```diff
@@ -0,0 +1,251 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::ErrorGuaranteed;
+use rustc_hir::{def::DefKind, def_id::LocalDefId};
+use rustc_middle::ty::util::{CheckRegions, NotUniqueParam};
+use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitableExt};
+use rustc_middle::ty::{TypeSuperVisitable, TypeVisitable, TypeVisitor};
+use rustc_span::{Span, DUMMY_SP};
+use rustc_type_ir::AliasKind;
+use std::ops::ControlFlow;
+
+use crate::errors::{DuplicateArg, LateBoundInOpaque, NotParam};
+
+struct OpaqueTypeCollector<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    opaques: Vec<ty::Binder<'tcx, Ty<'tcx>>>,
```

Comment:
All of these opaques are bound to the same bound vars, right?

Reply:
Oh, there's an optimization to clear bound vars if they're not referenced? I'd probably just do without that optimization, tbh...